pkg/router/controller/router_controller.go

```diff
@@ -227,6 +228,59 @@ func (c *RouterController) HandleEndpoints(eventType watch.EventType, obj interf
 	c.Commit()
 }
 
+// HandleEndpointSlice handles a single EndpointSlice event and refreshes the router backend.
+func (c *RouterController) HandleEndpointSlice(eventType watch.EventType, objMeta metav1.ObjectMeta, items []discoveryv1beta1.EndpointSlice) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var subsets []kapi.EndpointSubset
+
+	for i := range items {
+		var ports []kapi.EndpointPort
+		var addresses []kapi.EndpointAddress
+
+		for j := range items[i].Endpoints {
+			for k := range items[i].Endpoints[j].Addresses {
+				epa := kapi.EndpointAddress{
+					IP:        items[i].Endpoints[j].Addresses[k],
+					TargetRef: items[i].Endpoints[j].TargetRef,
+				}
+				if items[i].Endpoints[j].Hostname != nil {
+					epa.Hostname = *items[i].Endpoints[j].Hostname
+				}
+				addresses = append(addresses, epa)
+			}
+		}
+
+		for j := range items[i].Ports {
+			ports = append(ports, convertEndpointPort(items[i].Ports[j]))
+		}
+
+		subsets = append(subsets, kapi.EndpointSubset{
+			Addresses: addresses,
+			Ports:     ports,
+		})
+	}
+
+	endpoints := &kapi.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            objMeta.Name,
+			Namespace:       objMeta.Namespace,
+			Labels:          objMeta.Labels,
+			Annotations:     objMeta.Annotations,
+			OwnerReferences: objMeta.OwnerReferences,
+			ClusterName:     objMeta.ClusterName,
+		},
+		Subsets: subsets,
+	}
+
+	c.RecordNamespaceEndpoints(eventType, endpoints)
+	if err := c.Plugin.HandleEndpoints(eventType, endpoints); err != nil {
```

Comment:
I think simply passing along `eventType` is incorrect.  What if the slice was deleted, but another slice for the same service exists?

Looks like `RecordNamespaceEndpoints` and all `HandleEndpoints` implementations treat `watch.Modified` and `watch.Added` the same, so we can conflate `watch.Modified` and `watch.Added` here.  Something like the following could suffice:

```go
	if len(items) == 0 {
		eventType = watch.Deleted
	} else {
		eventType = watch.Modified
	}
```